File: flutter/hydroponics_framework/lib/datamodels/installation/installation_model.dart
import 'package:hydroponics_framework/datamodels/installation/physical_node_model.dart';
import 'package:hydroponics_framework/datamodels/json_model.dart';

class InstallationModel extends JsonModel {
  final String installationId;
  final String implementationId;
  final String name;
  final String description;
  final List<PhysicalNodeModel> physicalNodes;

  InstallationModel({
    required this.installationId,
    required this.implementationId,
    required this.name,
    required this.description,
    required this.physicalNodes
  }) : super();

  static InstallationModel fromJson(Map<String, dynamic> json) =>
      InstallationModel(
        installationId : json["installationId"],
        implementationId : json["implementationId"],
        name : json["name"],
        description : json["description"],
        physicalNodes : List<Map<String,dynamic>>.from(json["physicalNodes"])
            .map((p) => PhysicalNodeModel.fromJson(p))
            .toList()
  );


  Map<String, dynamic> toJson() {
    Map<String, dynamic> map = {};
    map["installationId"] = installationId;
    map["implementationId"] = implementationId;
    map["name"] = name;
    map["description"] = description;
    map["physicalNodes"] = physicalNodes.map((pn) => pn.toJson()).toList();
    return map;
  }
}

File: kotlin/hydroponics-host/src/main/kotlin/io/greenglass/hydroponics.application/models/installation/InstallationModel.kt
package io.greenglass.hydroponics.application.models.installation

import org.neo4j.graphdb.*

import io.greenglass.host.application.neo4j.GraphNode
import io.greenglass.hydroponics.application.models.jobs.JobModel.Graph.CURRENTJOB
import io.greenglass.hydroponics.application.models.jobs.JobModel.Graph.jobLabel
import io.greenglass.hydroponics.application.models.implementation.ImplementationModel.Graph.implementationLabel
import io.greenglass.hydroponics.application.models.installation.PhysicalNodeModel.Graph.PHYSICAL_NODE
import io.greenglass.hydroponics.application.models.system.HydroponicsSystemModel


class InstallationModel(
    val installationId : String,
    val implementationId : String,
    val name : String,
    val description : String,
    val physicalNodes : List<PhysicalNodeModel>,
    val currentJobId : String?
) : GraphNode {

    constructor(node : Node) : this(
        node.getProperty("installationId") as String,
        node.getSingleRelationship(INSTALLATION, Direction.INCOMING)
            .startNode
            .getProperty("implementationId") as String,
        node.getProperty("name") as String,
        node.getProperty("description") as String,
        node.getRelationships(Direction.OUTGOING, PHYSICAL_NODE, )
            .map { r -> PhysicalNodeModel(r.endNode) },
        if(node.hasRelationship(Direction.OUTGOING, CURRENTJOB))
             node.getSingleRelationship(CURRENTJOB, Direction.OUTGOING).endNode.getProperty("jobId") as String
        else
            null
    )
    override fun toNode(tx: Transaction): Node {
        val node = tx.createNode(installationLabel)
        val implNode = checkNotNull(
            tx.findNode(
                implementationLabel, "implementationId", implementationId
            )
        )
        implNode.createRelationshipTo(node, INSTALLATION)
        node.setProperty("installationId", installationId)
        node.setProperty("name", name)
        node.setProperty("description", description)
        physicalNodes.forEach { pn -> node.createRelationshipTo(pn.toNode(tx), PHYSICAL_NODE) }
        if (currentJobId != null) {
            val jobNode = checkNotNull(tx.findNode(jobLabel, "jobId", currentJobId))
            node.createRelationshipTo(jobNode, CURRENTJOB)
        }
        return node
    }

    fun toView(tx: Transaction) : InstallationViewModel {
        val impl = tx.findNode(implementationLabel, "implementationId", implementationId)
        val sys =  impl.getSingleRelationship(HydroponicsSystemModel.SYSTEM, Direction.OUTGOING).endNode

        return InstallationViewModel(
            installationId = this.installationId,
            instanceName = this.name,
            implementationId = impl.getProperty("implementationId") as String,
            implementationName = impl.getProperty("name") as String,
            systemId = sys.getProperty("systemId") as String,
            systemName = sys.getProperty("name") as String,
        )
    }

    companion object Graph {
        val installationLabel = Label.label("Installation")
        val INSTALLATION: RelationshipType = RelationshipType.withName("INSTALLATION")
    }
}
